Email alerts configuration form

ryanb-gds opened the pull request:
I'm not especially happy with how this has turned out, felt like I spent a lot of time fighting with Active Model to get the behaviour right. Sort of wonder if we should have just stuck with a plain Hash object to represent the finder schemas, but I've already spent far too long on this to start again now.

There's still a bit of work outstanding to make sure that external email signups are handled correctly by the metadata summary card but I want to make that a separate PR as I intend to refactor the summary card to a ViewComponent, as we will need more logic than I care to put in an erb template.

Worth reading the commit messages, they hopefully explain some of the more odd-looking code.

Trello: https://trello.com/c/l9XfLavV

ryanb-gds committed:
Handle email alert form inputs

This was a bit fiddly. The conditional field feature in the GOV.UK radio button component does not prevent the inclusion of the field values in the form submission, because it only hides the conditional DOM using CSS.

We therefore need to filter out any params for fields that should have been excluded based on the email alerts condition in the controller.

We also have to reset the values from the original schema, as we want the values from the form to overwrite the schema values if they change. We use the update method callback to do this.

I took the opportunity to update the signup_copy using an update callback as well, which will ensure we don't get any strange behaviour if the document noun and signup copy are assigned in an order we're not expecting.

ryanb-gds committed:
Ensure finder schema is output to Zendesk correctly

Now that the finder schema is a model and not a hash, we need to call `to_json` on it to render it inside the proposed schema hidden input value.

The `to_json` method is defined in the ActiveModel::Serializers::JSON module. To use it, I had to redefine the finder schema attributes using ActiveModel::Attributes.

The attributes usage is a little bit peculiar in that you can't write to them from inside the model, you have to use `#assign_attributes` instead. I tried all of `attribute =`, `self.attribute =`, `@attributes[:attribute] =` and `self[:attribute] =` without success.

ryanb-gds commented on app/models/finder_schema.rb:
```diff
@@ -14,74 +20,73 @@ def self.load_from_schema(type)
-    @related = value.nil? ? nil : value.reject(&:empty?)
+  def override_signup_copy
+    if signup_copy.present?
+      _assign_attribute(:signup_copy, "You'll get an email each time a #{document_noun} is updated or a new #{document_noun} is published.")
```
If anyone can find a less ugly way of doing this, please holler

ChrisBAshton replied:
Worth specifying in the commit message the various things you already tried (`self.<attribute_name> =`, etc). From dev huddle, sounds like you've already tried everything I would have, so no suggestions from me!

ryanb-gds commented on spec/models/finder_schema_spec.rb:
```diff
@@ -25,16 +25,17 @@
     end
   end
 
-  describe "#base_path" do
```
I decided to drop a good number of these tests as now that we are using Active Model attributes it felt like we were testing Rails. I have retained tests for all of our own code.

ChrisBAshton commented on app/components/email_alert_fieldset_component.html.erb:
```diff
@@ -3,16 +3,29 @@
   heading_level: 2,
   heading_size: "m",
 } do %>
-  <%= render "govuk_publishing_components/components/radio", {
```
Do we not need this character?

ryanb-gds replied:
Good point, that's actually a sneaky bug fix. I forgot to separate it into it's own commit

ryanb-gds replied:
Separated it out

ChrisBAshton commented on app/components/email_alert_fieldset_component.rb:
```diff
@@ -2,4 +2,57 @@ class EmailAlertFieldsetComponent < ViewComponent::Base
+
+    # Do not render the title prefix input if the value is a hash because we don't want to override the existing the existing
+    # values in such cases. We are going to revisit this later, but for now we only allow users to override this setting
+    # for finders with a single string value.
```
Would be nice to link to a Trello card in this comment (even if it's only a placeholder card for now).

ChrisBAshton commented on app/components/email_alert_fieldset_component.rb:
```diff
@@ -2,4 +2,57 @@ class EmailAlertFieldsetComponent < ViewComponent::Base
+        items: [
+          {
+            label: "Make changes to this filter criteria (The development team will reach out to you to discuss the requirements to allow users to sign up by specific filter criteria)",
+            value: "CHANGE_REQUESTED",
```
Nice.

ChrisBAshton commented on spec/components/email_alert_fieldset_component_spec.rb:
```diff
@@ -15,11 +15,69 @@
+    expect(page).to have_field("subscription_list_title_prefix", with: schema.subscription_list_title_prefix)
+  end
+
+  it "does not render the email subscription topic if the all_content value is checked but the current schema value is a hash" do
```
Worth adding a similar comment to what you added earlier, to imply the test is temporary and can be deleted one day?

ChrisBAshton commented on app/controllers/admin_controller.rb:
> The conditional field feature in the GOV.UK radio button component does not prevent the inclusion of the field values in the form submission, because it only hides the conditional DOM using CSS.

That's a shame - feels like a missing feature of the [checkbox](https://components.publishing.service.gov.uk/component-guide/checkboxes) component. Might be worth raising an issue on govuk-frontend? I [couldn't find an existing one](https://github.com/alphagov/govuk-frontend/issues?q=is%3Aissue+checkbox)

ryanb-gds replied:
I think this is more a feature of browser behaviour. The frontend team could remove the fields from the DOM via Javascript, but I suspect that might impact accessibility.

ChrisBAshton commented on app/models/finder_schema.rb:
```diff
@@ -2,6 +2,12 @@ class FinderSchema
+
+  define_model_callbacks :update
+
+  before_update :reset_email_alerts
```
Blocker: wouldn't this also reset the email alerts when we _don't_ want to reset them?
I.e. from the "edit metadata" or "edit filters/facets" forms?

ryanb-gds replied:
Yes - those other forms don't exist yet do they?

If/when they do we'll have to find another approach. Two spring to mind:

1. exclude these attributes when copying the attributes from the old schema into the proposed one. I didn't do this at first because I didn't want to put more logic into the controller. I don't feel strongly about that though
2. create some kind of update context option along the lines of Rails validation contexts to control which callbacks are applied

Any thoughts? There are probably other options as well

ChrisBAshton replied:
The first form already exists - see https://specialist-publisher.publishing.service.gov.uk/admin/metadata/aaib-reports

Of the two proposals, the second sounds best IMO.

ryanb-gds replied:
That's the same form I'm working on though (you'll notice the Email Alerts section with the legend in the wrong place at the bottom 🤦 )

Appreciate we'll have to do the work anyway when we come to filters and facets - happy to do it now if you'd prefer

ChrisBAshton replied:
Oh! I could have sworn we'd decided to make the email-signup-editing its own form. I guess not (or the designs have since been updated!).

Less of a reason to do it now if it's not going to break any existing forms. Just so long as we remember to bear it in mind for the facets form 👍 or you could do it now to reduce the risk of forgetting it.

ryanb-gds replied:
I suggest we make it a later if that's ok - no guarantees that we're going to get to filters and facets this quarter and I think it's going to require quite a lot of change to the finder schema model anyway, so this may need a rethink

ChrisBAshton reviewed (commented):
Looks good - I have one blocking comment, probably just me misunderstanding though. 🤞

ryanb-gds committed:
Fix rendering order bug in email alert fieldset component

Using `<%=` was rendering the radio component immediately to the DOM, resulting in the legend appearing underneath instead of above.

ryanb-gds committed:
Render fields for email alert signup options

Filtered email signups will not be handled via the Specialist Publisher admin form and therefore just submit a "CHANGE_REQUESTED" value for the email filter attribute.

Developers will contact the department offline to request more information.
